### ISSUE TITLE
Builder class not public even with BuilderVisibility.PUBLIC

### ISSUE THREAD

**Issue by ocrease** [immutables/immutables#628]
I'm using the following Style to create a package private implementation and a public builder:

```java
@Value.Style(
    visibility = Value.Style.ImplementationVisibility.PACKAGE,
    builderVisibility = Value.Style.BuilderVisibility.PUBLIC)
```

But in the generated classes both the implementation and builder are package private. I'll admit this is a workaround for a bug in Eclipse where the following setup fails when run from Eclipse (works fine from the command line).

```java
package immutables;

import org.immutables.value.Value;

@Value.Style(
    visibility = Value.Style.ImplementationVisibility.PACKAGE,
    builderVisibility = Value.Style.BuilderVisibility.PUBLIC)
@Value.Immutable
public abstract class Container
{
    public abstract int value();

    public static class Builder extends ImmutableContainer.Builder
    {
    }
}
```

```java
package immutables.test;

import java.util.function.Consumer;
import org.junit.Test;
import immutables.Container;

public class TestContainer
{
    @Test
    public void test()
    {
        Container.Builder container = new Container.Builder();
        invoke(container::value, 1);
    }
    
    private static void invoke(Consumer<Integer> method, int val) {
        System.out.println("Invoke - "+val);
        method.accept(val);
    }
}
```

When run in Eclipse you get the following runtime exception

```
java.lang.BootstrapMethodError: call site initialization exception
	at java.lang.invoke.CallSite.makeSite(CallSite.java:341)
	at java.lang.invoke.MethodHandleNatives.linkCallSiteImpl(MethodHandleNatives.java:307)
	at java.lang.invoke.MethodHandleNatives.linkCallSite(MethodHandleNatives.java:297)
	at immutables.test.TestContainer.test(TestContainer.java:13)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
Caused by: java.lang.IllegalArgumentException: java.lang.IllegalAccessException: class is not public: immutables.ImmutableContainer$Builder.value(int)Builder/invokeSpecial, from immutables.test.TestContainer
	at java.lang.invoke.MethodHandles$Lookup.revealDirect(MethodHandles.java:1360)
	at java.lang.invoke.AbstractValidatingLambdaMetafactory.<init>(AbstractValidatingLambdaMetafactory.java:131)
	at java.lang.invoke.InnerClassLambdaMetafactory.<init>(InnerClassLambdaMetafactory.java:155)
	at java.lang.invoke.LambdaMetafactory.metafactory(LambdaMetafactory.java:299)
	at java.lang.invoke.CallSite.makeSite(CallSite.java:302)
	... 26 more
Caused by: java.lang.IllegalAccessException: class is not public: immutables.ImmutableContainer$Builder.value(int)Builder/invokeSpecial, from immutables.test.TestContainer
	at java.lang.invoke.MemberName.makeAccessException(MemberName.java:850)
	at java.lang.invoke.MethodHandles$Lookup.checkAccess(MethodHandles.java:1536)
	at java.lang.invoke.MethodHandles$Lookup.revealDirect(MethodHandles.java:1357)
	... 30 more
```

Having done some investigation into this issue, if the ImmutableContainer.Builder was public, then this error would not happen. Clearly this is a bug in the eclipse compiler since it is not an issue when compiled with javac.

However, if I specify `builderVisibility = Value.Style.BuilderVisibility.PUBLIC` I would expect the generated builder to be public.

**Comment by ocrease** [comment]
FYI the bug in eclipse is https://bugs.eclipse.org/bugs/show_bug.cgi?id=517299

**Comment by elucash** [comment]
I've tried this with relatively new milestone Eclipse 4.7M7 and I see no problem compiling and running the example. Anyway, I've revised visibility restriction on nested builder, so if this (making public nested builder class inside package private implementation class) will solve the problem, then you can try just released v2.5.2.

**Comment by ocrease** [comment]
Thanks, works as expected on v2.5.2
